### ISSUE TITLE
Running unit tests on host PC *and* embedded target platform

### ISSUE THREAD

**Issue by 6arms1leg** [ThrowTheSwitch/Ceedling#448]
I am in the tricky situation that I need to run my unit tests on my development PC *and additionally* on the embedded target platform. The latter is a requirement from the customer.
I know that unit testing on the target is discouraged by the community (from [here](http://www.throwtheswitch.org/build/which) and other posts) but, so far, the customer insists. Using a simulator is also not an option for the customer.

In his book, "Test-Driven Development for Embedded C", J. Grenning advises to run unit tests on both, the host PC (all the time) and the embedded target (occasionally). Is that just the academical, ideal view or is there another reason for this disagreement?

So disregarding the fact that it is considered bad practice, I am currently evaluating such a setup with Ceedling.
My main concerns are:

1. **Registers:** One of the arguments against running unit tests on the target is that there is no control over the registers. But couldn't I just run the unit tests in pretty much the same way as I do on my host PC by using the same memory map? (This way all registers would still be simple variables.)
1. **Two test compilers:** Is there a way to define two `:test_compiler:` options (one for native and one for target builds) in the `project.yml` config file?
1. **Only one test executable:** Each test file generates one executable. Flashing that many executables each time I run all unit tests would be hard on the targets storage and take a long time. Is there an option to combine *all* unit tests into one executable to mitigate this problem (reduced write cycles)?
1. **Compile and flash automation**: Can Ceedling be used to assist in the flash process (s. th. more automated than using the `:build_only:` option and then flash manually)? Can the `:test_fixture:` option be used for that?

Currently, I foresee the following coarse (simplified) steps:

* Routing `$stdout` to UART, as described [here (somewhere)](https://github.com/ThrowTheSwitch/Ceedling/blob/master/docs/CeedlingPacket.md#the-almighty-project-configuration-file-in-glorious-yaml),
* somehow configure Ceedling to compile unit tests for the host PC and the target platform,
* somehow "bundle" all unit tests into only one executable,
* connect the target platform to the host PC (including UART),
* flash the (single) unit test executable and collect results via UART (occasionally),

and 

* also compile and run unit tests on the host PC in the "normal" way (all the time).

Is my understanding and approach too simplified? Am I missing something essential here?

Many many thanks for your help!

---

**tl;dr:** My goal is to write a (single) set of unit tests for my release code and run this (same) set of unit tests on both, the host PC and the embedded target platform, and both times faking registers with variables.

**Comment by mvandervoord** [comment]
You are on the right track. Setting up a project to use a target processor is almost identical to getting it working with a simulator, except you need to somehow ferry data back and forth (as you've already identified).

The only thing different than what you are saying is that ceedling will still build and deploy each test module separately, instead of all the unit tests being splatted into a single unit. This allows you to continue to use mocks, etc without having to worry about multiple definitions.

The last piece of the puzzle that you're missing is that you want to create your main `project.yml` file for your host PC builds... and then a second project file like `target.yml` for the target. This second file will just have the changes needed to the tools definitions, etc.

Then, to run tests on the target, it's just a matter of specifying you want to run your tests there:

```
ceedling project:target test:all
```

**Comment by 6arms1leg** [comment]
Many thanks for your help and that you took the time to read my rather long post!

> The only thing different than what you are saying is that ceedling will still build and deploy each test module separately, instead of all the unit tests being splatted into a single unit.

(This refers to my 3. concern in my initial post: **Only one test executable**)
So this is a fixed constraint (because of the issue with multiple definitions)?
As I wrote in my 3. concern above, I would like to reduce the write cycles to my target and the overall test time. Suppose I have *N* modules (~ `*.c` files) and a unit test file for each. That would mean I had to flash the MCU *N* times every time I run the unit tests on the target. It is not a blocker but it isn't ideal either.

> The last piece of the puzzle that you're missing is that you want to create your main project.yml file for your host PC builds... and then a second project file like target.yml for the target. This second file will just have the changes needed to the tools definitions, etc.

(This refers to my 2. concern in my initial post: **Two test compilers**)
You mean essentially having two almost identical Ceedling `*.yml` files that only differ in the `:tools:` section? Is there a mechanism in Ceedling to avoid the duplication in the files like an conditional `include`?
This way I don't have to make identical changes in two files. Something like a `main.yml` that either includes `host.yml` or `target.yml` in the `:tools:` section.

Sorry for my perseverance but could you please also comment on my 1. (**Registers**) and 4. (**Compile and flash automation**) concerns/questions in my initial post? It would really help me out a lot.

Many many thanks for your time!

**Comment by mvandervoord** [comment]
For concern#3, this is built into the way that Ceedling works. It tests a unit at a time to be able to separate out dependencies on other units. If we were working with a language that was easier to swap functions in and out, we could pull this off, but not C. 

It IS possible to write all of your code in a way that you can test it all from a single test executable. This would require you to use a lot of function pointers. When you want the mock, you'd assign to the mock. When you want the real version, you'd assign the function pointer to the real version. It would require a mockable header that had a slightly different name for every function than the original name, to keep from getting collisions. Honestly, I think it'd be more trouble than it's worth.

For concern#2, I think we have a small misunderstanding. In the second yaml file, the ONLY thing you need to define is the differences... namely the tools. Ceedling will start by loading the main project.yml file, and then will load the second over the top of it, replacing any parts that are specified and are different. In this way, you have a single project yaml file with most of your definition for your project... and then 1+ additional project files that just have variations. Think of them as patches that you can apply on the fly. ;)

For concern#1, I don't see any reason you can't do exactly as you say and work with additional structures. It's likely to take up quite a bit of memory (which depending on your target, might be a big deal) but otherwise would work fine.

For concern#4: This is what test_fixture is intended for. Your job would be to write a script or batch file or whatever to automate your flash process. Then you set that call as the test_fixture executable and everything should run. Depending on your build and flash process, this can be challenging, but if you can call things from the command line, ceedling can use it.

**Comment by 6arms1leg** [comment]
Again, thanks for your help!
That perfectly answers all my questions.

Your testing framework is amazing and I really appreciate that you take the time to help people out--like in this issue. I tried to show my gratitude by booking your online training and buying your book last year, which, by the way, are both great!

Keep up the good work!

**Comment by mvandervoord** [comment]
Thank you! I really appreciate that!  :)

**Comment by Sezginnn** [comment]
What is the latest situation? does the ceedling work on the target platform?

**Comment by Sezginnn** [comment]
What is the latest situation? does the ceedling work on the target platform?

**Comment by Letme** [comment]
As mentioned in another thread: it does not have to work on target platform, it just needs to upload (flash) and read data from it. But if target platform has linux and ruby on, then it will run without a problem directly on target platform.
